File: src/main/java/incjc/ProcHelpers.java
package incjc;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.StringWriter;
import java.io.Writer;
import java.nio.charset.StandardCharsets;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Optional;
import org.apache.commons.io.IOUtils;

public abstract class ProcHelpers {

    public static String getJdkExecutable(String name) {
        String jdkHome = Optional.ofNullable(System.getenv("JDK_HOME")).orElse(System.getenv("JAVA_HOME"));
        return jdkHome != null ? Paths.get(jdkHome, "bin", name).toString() : name;
    }

    public static String getProcessOutput(String[] commandLine) {
        Writer output = new StringWriter();
        try {
            Process p = Runtime.getRuntime().exec(commandLine);
            Thread inputPump = inputStreamPump(p.getInputStream(), output);
            inputPump.start();
            int retval = p.waitFor();
            inputPump.join();
            if (retval != 0) {
                throw new RuntimeException(String.format("Nonzero return value of %d was returned by %s",
                    retval, String.join(" ", Arrays.asList(commandLine))));
            }
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }

        return output.toString();
    }

    public static Thread inputStreamPump(InputStream from, OutputStream to) {
        return new Thread(() -> {
            try {
                IOUtils.copy(from, to);
            } catch (IOException e) {
                e.printStackTrace(System.err);
            }
        });
    }

    public static Thread inputStreamPump(InputStream from, Writer to) {
        return new Thread(() -> {
            try {
                IOUtils.copy(from, to, StandardCharsets.US_ASCII);
            } catch (IOException e) {
                e.printStackTrace(System.err);
            }
        });
    }
}

File: src/main/kotlin/ProcHelpers.kt
package incjc

import org.apache.commons.io.IOUtils
import java.io.*
import java.nio.charset.StandardCharsets
import java.nio.file.Paths

fun getJdkExecutable(name: String): String {
    val jdkHome = System.getenv("JDK_HOME") ?: System.getenv("JAVA_HOME")
    return if (jdkHome != null) Paths.get(jdkHome, "bin", name).toString() else name
}

fun getProcessOutput(commandLine: Array<String?>): String {
    val output = StringWriter()
    try {
        val p = Runtime.getRuntime().exec(commandLine)
        val inputPump = inputStreamPump(p.inputStream, output)
        inputPump.start()
        val retval = p.waitFor()
        inputPump.join()
        if (retval != 0) {
            val cmd = commandLine.joinToString(" ")
            throw RuntimeException("Nonzero return value of $retval was returned by $cmd")
        }
    } catch (e: IOException) {
        throw RuntimeException(e)
    } catch (e: InterruptedException) {
        throw RuntimeException(e)
    }
    return output.toString()
}

fun inputStreamPump(from: InputStream, to: OutputStream): Thread {
    return Thread {
        try {
            IOUtils.copy(from, to)
        } catch (e: IOException) {
            e.printStackTrace(System.err)
        }
    }
}

fun inputStreamPump(from: InputStream?, to: Writer?): Thread {
    return Thread {
        try {
            IOUtils.copy(from, to, StandardCharsets.US_ASCII)
        } catch (e: IOException) {
            e.printStackTrace(System.err)
        }
    }
}
